### ISSUE TITLE
[struct,list,map]Type markers on types are ignored

### ISSUE THREAD

**Issue by jan--f** [kubernetes-sigs/controller-tools#676]
Let me preface this by admitting that I'm not exactly sure if this is a bug or expected behavior. I might also just not use the tooling correctly to get the outcome I want.

Server-side-apply makes use of markers (like `structType=atomic`) to merge resources and partial updates to them. controller-tools already [respects those markers on fields](https://github.com/kubernetes-sigs/controller-tools/blob/a260f1391b0f722e07ae04fe87e23e32bc1a1ec7/pkg/crd/testdata/cronjob_types.go#L151) and produces the [respective yaml annotations](https://github.com/kubernetes-sigs/controller-tools/blob/a260f1391b0f722e07ae04fe87e23e32bc1a1ec7/pkg/crd/testdata/testdata.kubebuilder.io_cronjobs.yaml#L7321).

When I mark the type `NestedObject` though the yaml annotation is not generated. To test this one can apply the following patch and just run `make`, which will now produce an error.
``` patch
diff --git a/pkg/crd/testdata/cronjob_types.go b/pkg/crd/testdata/cronjob_types.go
index 66222f6e..5d28063d 100644
--- a/pkg/crd/testdata/cronjob_types.go
+++ b/pkg/crd/testdata/cronjob_types.go
@@ -148,7 +148,6 @@ type CronJobSpec struct {
        MapOfInfo map[string][]byte `json:"mapOfInfo"`

        // A struct that can only be entirely replaced
-       // +structType=atomic
        StructWithSeveralFields NestedObject `json:"structWithSeveralFields"`

        // This tests that type references are properly flattened
@@ -258,6 +257,7 @@ func (p *Preserved) MarshalJSON() ([]byte, error) {
        return json.Marshal(full)
 }

+// +structType=atomic
 type NestedObject struct {
        Foo string `json:"foo"`
        Bar bool   `json:"bar"`
```

Now I think the two variants I described should produce the same results. For example if I use `metav1.LabelSelector` in a type declaration (from which a CRD manifest is generated), I currently have to explicitly add the marker to the field of that type. However the [type already has the marker that I want](https://github.com/kubernetes/apimachinery/blob/97e5df2d0258ac077093867fabf508c1fc31f53b/pkg/apis/meta/v1/types.go#L1181) and I think there's an argument to be made for letting the type define this behavior.

The current behavior produces surprising results when existing controller code wants to adopt a server-side-apply workflow (over a read-modify-write workflow) where multiple entries in a LabelSelector field can have a different manager each. See for example https://github.com/rhobs/monitoring-stack-operator/pull/121.

**Comment by jan--f** [comment]
Thanks! I'll try to confirm this and then close this here accordingly.

**Comment by jan--f** [comment]
@robbie-demuth this has indeed been resolved.
I only have one minor follow up fix, proposed in #704.
